### DISCUSSION TITLE
Improving Central Orchestrator Agent Classification

### DISCUSSION THREAD

**Discussion by anjineyulutv** [PeoplePlusAI/Sukoon#3] [Ideas]
### **Scenario Overview**

In our Multi-Agent System, the Central Orchestrator Agent plays a crucial role in analyzing incoming queries and delegating tasks to appropriate specialized agents. Recently, we've observed a concerning trend of misclassifications, leading to suboptimal task routing and inefficient use of our specialized agents.

### **The Problem**

The Central Orchestrator Agent is currently experiencing difficulties in accurately categorizing complex, multifaceted queries. This misclassification results in:

1.Inappropriate agent selection for certain tasks.

2.Inefficient use of system resources.

3.Longer processing times for queries.

4.Potential errors in final outputs due to mismatched expertise.

These issues are particularly pronounced when dealing with queries that span multiple domains or require nuanced understanding of context.

### **Guiding Questions**

To help you approach this task systematically, consider the following questions:

1.What patterns do you observe in the types of queries that are most frequently misclassified?

2.How does the current classification algorithm work? What are its strengths and limitations?

3.Are there any biases in the training data that might be affecting the agent's decision-making?

4.How can we incorporate context and multi-domain understanding into the classification process?

5.What additional metadata or information could be helpful in improving classification accuracy?

6.How might we implement a confidence score for the agent's classifications, and how could this be used in the routing process?

7.Are there opportunities for implementing a feedback loop to continuously improve classification accuracy over time?

### **Expected Deliverables**

1.A detailed analysis of the current classification system and its shortcomings.

2 Proposed improvements to the classification algorithm, including any necessary changes to the agent's architecture.

3.A plan for gathering and incorporating more comprehensive metadata for incoming queries.

4.Suggestions for implementing a multi-stage classification process for complex queries.

5.A proposed method for measuring and monitoring classification accuracy over time.

### Additional Considerations

1.Consider the trade-offs between classification accuracy and processing speed. How can we balance these factors?

2.Think about how we can make the classification process more transparent and explainable, both for system audits and for user understanding.

3.Consider the scalability of your proposed solutions as the system grows and handles an increasing variety of query types.

**Comment by luv-singh-ai** [comment]
Adding to above,
1. How do we improve the reasoning capabilities of LLMs particularly with multiple tools?
2. re: question 1 
Many times I've seen that agents will not make tool calls even when it's required. or make it half of the times, and fail at other times. this is more prominent when they don't get which arguments to be supplied to the function call. eg ignoring required parameters 
3. How do we improve the LLM capability to commit important info in long term memory?

**Reply by anjineyulutv** [reply]
Good points!!!Semantic cache will be a top priority for capping compute too.

**Comment by Dhanu-shikaa** [comment]
An idea for the above,

To avoid query misclassification, we can use POS tagging to count verbs or noun. Based on the verb or noun count, employing unigram for simple queries, bigrams for moderately complex ones, or n-grams for more intricate requests. This approach tailors the classification method to the query's complexity, potentially improving accuracy.

**Reply by anjineyulutv** [reply]
As a next step let's understand the effect of each part of speech on accuracy.

**Comment by Maha256** [comment]
To minimize the long processing times of COAs, we can implement Asynchronous Task Processing, enabling the COA to handle tasks based on events or triggers (Event-Driven Architecture). This approach improves responsiveness. Incorporating user context and geographical information as metadata can reduce query misclassification.

**Comment by HarishNarasimhanK0135** [comment]
In order to find patterns of a misclassified queries, 
we can collect all queries that have been misclassified by the agent, and combine these queries by the type of specialized agent they were incorrectly assigned to. Then, we can apply clustering to identify patterns in the queries that were incorrectly routed. This may help in understanding why the Central Orchestrator agent misclassifies these queries.

**Comment by vigneshmj1997** [comment]
Guys,
Let's first work on creating a benchmark ,Then later create a baseline score with a basic code and then work on improving the benchmark later on 
What say !!

**Comment by chetan-k-s** [comment]
An idea for the above,

To implement a confidence score for the agent's classification, we can quantify the confidence score as the product of the probability of the query being classified correctly and the actual classification. We can plot F1 score-Confidence, Precision-Confidence and Recall-Confidence graphs and determine the optimal confidence score. This way the number of misclassifications can be reduced.

**Reply by anjineyulutv** [reply]
Please try to simulate the ideas with multiple LLMs each of Specialized agents

**Comment by anjineyulutv** [comment]
We can start using a simple algorithm and predictions having higher threshold values can be directly taken for granted and as predictions tend towards 0.5 (nuetrality of classification) we can use more complex algorithms for inferencing.We can have a auto-tuner to select how many models and for what range of thresholds we need to select what models.

**Comment by anjineyulutv** [comment]
@luv-singh-ai .Do u have any benchmarks in place?

**Reply by luv-singh-ai** [reply]
We can have Per-Agent Confusion matrix( Precision, Recall, and F1 Score) which will help identify which specialised agents are most affected by misclassification.

**Reply by anjineyulutv** [reply]
Can u provide the dataset.We are working on synthetically generated data

**Reply by luv-singh-ai** [reply]
Sure, Anjineyulu. These were prompts that were tried out by IITk students to the chatbot and you can see some of them were misclassified. eg motivational_agent was not invoked, instead suicide prevention was invoked. though this is a small dataset, we can start from this one. and this expands, we can can add more edge cases here.
Here's the link - https://docs.google.com/spreadsheets/d/1QIeV1pa8z0xCGhAjSWoe0yBH2zVmrA0jc0j3HOztNNI/edit?usp=sharing

**Reply by luv-singh-ai** [reply]
List of all agents and prompts are in this notebook- https://github.com/PeoplePlusAI/Sukoon/blob/icg/Sukoon_added_agents.ipynb

**Comment by anjineyulutv** [comment]
As next steps a dataset curation will happen and quantitative evaluation of diversification of data will be formulated to make data better for model tarining
